File: Core/Kernel/KRWEngine.m
#import "KRWEngine.h"
#import "darksword.h"
#import "Logger.h"

@implementation KRWEngine

+ (instancetype)shared {
    static KRWEngine *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[self alloc] init];
    });
    return instance;
}

- (uint64_t)kread64:(uint64_t)addr {
    return ds_kread64(addr);
}
- (uint32_t)kread32:(uint64_t)addr {
    return ds_kread32(addr);
}
- (void)kwrite64:(uint64_t)addr value:(uint64_t)value {
    ds_kwrite64(addr, value);
}
- (void)kwrite32:(uint64_t)addr value:(uint32_t)value {
    ds_kwrite32(addr, value);
}
- (void)kwrite8:(uint64_t)addr value:(uint8_t)value {
    ds_kwrite8(addr, value);
}

@end

File: Core/Kernel/KRWEngine.swift
import Foundation

class KRWEngine: @unchecked Sendable {
    static let shared = KRWEngine()

    func kread64(_ addr: UInt64) -> UInt64 {
        return ds_kread64(addr)
    }

    func kread32(_ addr: UInt64) -> UInt32 {
        return ds_kread32(addr)
    }

    func kwrite64(_ addr: UInt64, _ value: UInt64) {
        ds_kwrite64(addr, value)
    }

    func kwrite32(_ addr: UInt64, _ value: UInt32) {
        ds_kwrite32(addr, value)
    }

    func kwrite8(_ addr: UInt64, _ value: UInt8) {
        ds_kwrite8(addr, value)
    }
}
